File: VyacheslavBuchin/term2/block2/ArrayList/src/test/java/collection/ArrayListTest.java
package collection;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

class ArrayListTest {

	private ArrayList<Integer> testList;
	private java.util.ArrayList<Integer> referenceList;
	private static final int[] values = { 0, -1, 1, Integer.MAX_VALUE, Integer.MIN_VALUE };
	private static final int[] inBoundsIndexes = { 0, 50, 99 };
	private static final int[] outOfBoundsIndexes = { -1, 101, 1000 };

	private static Stream<Arguments> values() {
		return Arrays.stream(values).mapToObj(Arguments::of);
	}

	private static Stream<Arguments> inBoundsIndexes() {
		return Arrays.stream(inBoundsIndexes).mapToObj(Arguments::of);
	}

	private static Stream<Arguments> outOfBoundsIndexes() {
		return Arrays.stream(outOfBoundsIndexes).mapToObj(Arguments::of);
	}

	private static Stream<Arguments> inBoundsIndexedValues() {
		var argList = new java.util.ArrayList<Arguments>();
		for (var index : inBoundsIndexes) {
			for (var value : values) {
				argList.add(Arguments.of(index, value));
			}
		}
		return argList.stream();
	}

	private static Stream<Arguments> outOfBoundsIndexedValues() {
		var a = new java.util.ArrayList<Arguments>();
		for (var index : outOfBoundsIndexes) {
			for (var value : values) {
				a.add(Arguments.of(index, value));
			}
		}
		return a.stream();
	}

	private static Stream<Arguments> rangesInBounds() {
		return Stream.of(
				Arguments.of(0, 100),
				Arguments.of(5, 15)
		);
	}

	private static Stream<Arguments> rangesOutOfBounds() {
		return Stream.of(
				Arguments.of(0, 115),
				Arguments.of(-3, 10),
				Arguments.of(228, 1337)
		);
	}

	@BeforeEach
	void setUp() {
		testList = new ArrayList<>();
		referenceList = new java.util.ArrayList<>();
		for (int i = 0; i < 100; i++) {
			testList.add(i);
			referenceList.add(i);
		}
	}

	@Test
	void sizeTest() {
		assertEquals(referenceList.size(), testList.size());
	}

	@Test
	void isEmptyTest() {
		assertEquals(referenceList.isEmpty(), testList.isEmpty());
	}

	@ParameterizedTest
	@MethodSource("values")
	void containsTest(int value) {
		assertEquals(referenceList.contains(value), testList.contains(value));
	}

	@Test
	void toArrayTest() {
		assertArrayEquals(referenceList.toArray(), testList.toArray());
	}

	@ParameterizedTest
	@MethodSource("inBoundsIndexedValues")
	void addTestWithIndexesInBounds(int index, int value) {
		referenceList.add(index, value);
		testList.add(index, value);
		assertArrayEquals(referenceList.toArray(), testList.toArray());
	}

	@ParameterizedTest
	@MethodSource("outOfBoundsIndexedValues")
	void addTestWithIndexesOutOfBounds(int index, int value) {
		assertThrows(IndexOutOfBoundsException.class, () -> referenceList.add(index, value));
		assertThrows(IndexOutOfBoundsException.class, () -> testList.add(index, value));
	}

	@ParameterizedTest
	@MethodSource("inBoundsIndexes")
	void removeInBoundsIndexedTest(int index) {
		referenceList.remove(index);
		testList.remove(index);
		assertArrayEquals(referenceList.toArray(), testList.toArray());
	}

	@ParameterizedTest
	@MethodSource("outOfBoundsIndexes")
	void removeOutOfIndexedTest(int index) {
		assertThrows(IndexOutOfBoundsException.class, () -> referenceList.remove(index));
		assertThrows(IndexOutOfBoundsException.class, () -> testList.remove(index));
	}

	@ParameterizedTest
	@MethodSource("values")
	void removeObjectTest(int value) {
		assertEquals(referenceList.remove((Object) value), testList.remove((Object) value));
		assertArrayEquals(referenceList.toArray(), testList.toArray());
	}

	@Test
	void containsAllTest() {
		assertTrue(testList.containsAll(testList));
		assertTrue(testList.containsAll(referenceList));
		assertTrue(referenceList.containsAll(testList));
	}

	@Test
	void retainAllTest() {
		var tempArray = testList.toArray();
		testList.retainAll(referenceList);
		assertArrayEquals(tempArray, testList.toArray());
	}

	@ParameterizedTest
	@MethodSource("inBoundsIndexes")
	void getInBoundsTest(int index) {
		assertEquals(referenceList.get(index), testList.get(index));
	}

	@ParameterizedTest
	@MethodSource("outOfBoundsIndexes")
	void getOutOfBoundsTest(int index) {
		assertThrows(IndexOutOfBoundsException.class, () -> referenceList.get(index));
		assertThrows(IndexOutOfBoundsException.class, () -> testList.get(index));
	}

	@ParameterizedTest
	@MethodSource("inBoundsIndexedValues")
	void setInBoundsTest(int index, int value) {
		referenceList.set(index, value);
		testList.set(index, value);
		assertArrayEquals(referenceList.toArray(), testList.toArray());
	}

	@ParameterizedTest
	@MethodSource("outOfBoundsIndexedValues")
	void setOutOfBoundsTest(int index, int value) {
		assertThrows(IndexOutOfBoundsException.class, () -> referenceList.set(index, value));
		assertThrows(IndexOutOfBoundsException.class, () -> testList.set(index, value));
	}

	@Test
	void addAllTest() {
		var appendingList = List.of(228, 1337, 5553535);
		referenceList.addAll(appendingList);
		testList.addAll(appendingList);
		assertArrayEquals(referenceList.toArray(), testList.toArray());
	}

	@Test
	void removeAllTest() {
		var removingList = List.of(228, 1337, 5553535, 1, 2, 3, 99);
		referenceList.removeAll(removingList);
		testList.removeAll(removingList);
		assertArrayEquals(referenceList.toArray(), testList.toArray());
	}

	@Test
	void toArrayTypedTest() {
		var referenceArray = new Number[referenceList.size()];
		var testArray = new Number[testList.size()];
		referenceList.toArray(referenceArray);
		testList.toArray(testArray);
		assertArrayEquals(referenceArray, testArray);
	}

	@ParameterizedTest
	@MethodSource("rangesInBounds")
	void subListInBoundsTest(int from, int to) {
		var referenceArray = referenceList.subList(from, to).toArray();
		var testArray = testList.subList(from, to).toArray();
		assertArrayEquals(referenceArray, testArray);
	}

	@ParameterizedTest
	@MethodSource("rangesOutOfBounds")
	void subListOutOfBoundsTest(int from, int to) {
		assertThrows(
				IndexOutOfBoundsException.class,
				() -> referenceList.subList(from, to)
		);
		assertThrows(
				IndexOutOfBoundsException.class,
				() -> testList.subList(from, to)
		);
	}
}

File: VyacheslavBuchin/term2/block2/ArrayListKotlin/src/test/kotlin/collection/ArrayListTest.kt
package collection

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import kotlin.test.assertTrue

internal class ArrayListTest {

	private val testList = ArrayList<Int>()
	private val referenceList = kotlin.collections.ArrayList<Int>()

	companion object {
		private val values = intArrayOf(0, -1, 1, Int.MAX_VALUE, Int.MIN_VALUE)
		private val inBoundsIndexes = intArrayOf(0, 50, 99)
		private val outOfBoundsIndexes = intArrayOf(-1, 101, 1000)

		@JvmStatic
		private fun values() = values.map { Arguments.of(it) }

		@JvmStatic
		private fun inBoundsIndexes() = inBoundsIndexes.map { Arguments.of(it) }

		@JvmStatic
		private fun outOfBoundsIndexes() = outOfBoundsIndexes.map { Arguments.of(it) }

		@JvmStatic
		private fun inBoundsIndexedValues(): List<Arguments> {
			val argList = mutableListOf<Arguments>()
			inBoundsIndexes.forEach { index ->
				values.forEach { value ->
					argList.add(Arguments.of(index, value))
				}
			}
			return argList
		}

		@JvmStatic
		private fun outOfBoundsIndexedValues(): List<Arguments> {
			val argList = mutableListOf<Arguments>()
			outOfBoundsIndexes.forEach { index ->
				values.forEach { value ->
					argList.add(Arguments.of(index, value))
				}
			}
			return argList
		}
	}
	@BeforeEach
	fun setUp() {
		for (i in 0..99) {
			testList.add(i)
			referenceList.add(i)
		}
	}

	@Test
	fun `size test`() {
		assertEquals(referenceList.size, testList.size)
	}

	@Test
	fun `isEmpty test`() {
		assertEquals(referenceList.isEmpty(), testList.isEmpty())
	}

	@ParameterizedTest
	@MethodSource("values")
	fun `contains test`(value: Int) {
		assertEquals(referenceList.contains(value), testList.contains(value))
	}

	@ParameterizedTest
	@MethodSource("inBoundsIndexedValues")
	fun `add should insert value in given index if index is in bounds`(index: Int, value: Int) {
		referenceList.add(index, value)
		testList.add(index, value)
		assertTrue(referenceList.isEqualTo(testList))
	}

	@ParameterizedTest
	@MethodSource("outOfBoundsIndexedValues")
	fun `add should throw exception if index is not in bounds`(index: Int, value: Int) {
		assertThrows(IndexOutOfBoundsException::class.java) { referenceList.add(index, value) }
		assertThrows(IndexOutOfBoundsException::class.java) { testList.add(index, value) }
	}

	@ParameterizedTest
	@MethodSource("inBoundsIndexes")
	fun `remove should remove value from given index if index is in bounds`(index: Int) {
		referenceList.removeAt(index)
		testList.remove(index)
		assertTrue(referenceList.isEqualTo(testList))
	}

	@ParameterizedTest
	@MethodSource("outOfBoundsIndexes")
	fun `remove should throw exception if index is not in bounds`(index: Int) {
		assertThrows(IndexOutOfBoundsException::class.java) { referenceList.removeAt(index) }
		assertThrows(IndexOutOfBoundsException::class.java) { testList.remove(index) }
	}

	@ParameterizedTest
	@MethodSource("values")
	fun `remove should remove given object if it is in list`(value: Int) {
		assertEquals(referenceList.remove(value as Any), testList.remove(value as Any))
		assertTrue(referenceList.isEqualTo(testList))
	}

	@ParameterizedTest
	@MethodSource("inBoundsIndexes")
	fun `get should return value stored in given index`(index: Int) {
		assertEquals(referenceList[index], testList[index])
	}

	@ParameterizedTest
	@MethodSource("outOfBoundsIndexes")
	fun `get should throw exception if given index is not in bounds`(index: Int) {
		assertThrows(IndexOutOfBoundsException::class.java) { referenceList[index] }
		assertThrows(IndexOutOfBoundsException::class.java) { testList[index] }
	}

	@ParameterizedTest
	@MethodSource("inBoundsIndexedValues")
	fun `set should change value by the given index`(index: Int, value: Int) {
		referenceList[index] = value
		testList[index] = value
		assertTrue(referenceList.isEqualTo(testList))
	}

	@ParameterizedTest
	@MethodSource("outOfBoundsIndexedValues")
	fun `set should throw exception if given index is not in bounds`(index: Int, value: Int) {
		assertThrows(IndexOutOfBoundsException::class.java) { referenceList[index] = value }
		assertThrows(IndexOutOfBoundsException::class.java) { testList[index] = value }
	}


}

fun kotlin.collections.ArrayList<Int>.isEqualTo(o: ArrayList<Int>): Boolean {
	this.forEachIndexed { index, element ->
		if (element != o[index])
			return false
	}
	return true
}